src/Features/Core/Portable/QuickInfo/QuickInfoContext.cs

```diff
@@ -14,30 +18,89 @@ internal sealed class QuickInfoContext
     {
         /// <summary>
         /// The document that quick info was requested within.
+        /// Can be <see langword="null"/> for quick info scenarios based off a <see cref="Compilation"/>,
+        /// where <see cref="SemanticModel"/> is guaranteed to be non-null.
         /// </summary>
-        public Document Document { get; }
+        public Document? Document { get; }
+
+        /// <summary>
+        /// The semantic model that quick info was requested within.
+        /// Can be <see langword="null"/> for quick info scenarios where <see cref="Document.SupportsSemanticModel"/> is <see langword="false"/>
+        /// for the <see cref="Document"/> whose quick info is being requested,
+        /// </summary>
+        public SemanticModel? SemanticModel { get; }
```

Comment:
the big issue here is that QI is a *cross language* concept.  But we're tying this tightly to C#/VB compilation model here.  I think a better approach instead would be if we could make the LSIF generator just produce Documents that wrap the compilation objects.  That way we could use all these APIs without an issue.

This also matters if we want to support things like LSIF for F#/TS/JS.

I *Really* cannot express how much i do not like this.  It strongly couples the featuers and the compilatino model again, and that's something we intentionally avoided entirely from the beginning here.

Reply:
> I think a better approach instead would be if we could make the LSIF generator just produce Documents that wrap the compilation objects

Tagging @jasonmalinowski. I believe I tried this approach in LSIF generator for synthesizing document/project for folding support as well, but that led to concerns that now we have a Document object being passed around that does not represent the true document in the solution (for example, if it was a linked document in real, our synthesized document would not be one). During that feature we concluded that we are not going to go the route of synthesizing document/project solutions from compilation, but instead refactor these services to be able to work for both kinds of inputs: 1. SemanticModel/Compilation for LSIF generation and 2. Document/Project for rest of IDE usage.

I think we need to finalize one approach and then use it across all LSIF features: either we synthesize documents in LSIF generator and use it across all services that need LSIF support OR we refactor these services to work in both modes: semantic model/compilation + document/workspace

Reply:
> OR we refactor these services to work in both modes: semantic model/compilation + document/workspace

I really dont' want this, and i think this is really an inappropriate coupling between a feature and C#/VB.  Even if it's optional, we're effectively saying: Every feature that might be lsif related needs to have a codepath for non-C#/VB and a specialzied codepath baked in for C#/VB.  That just screams wrongness *deeply* to me.  We have an entire language-neutral abstraction mechanism exactly for this purpose.  We're not using it for lsif and we end up here.  But i think we need to examine why we're nto using it.

Note: if we must avoid that abstraction, i would say we should have an entire separate C#/VB specific set of feature abstractions that these features can implement *as well*.  but we should not taint the core set of language neutral interfaces we already have.   :)

Reply:
Note that if we do this, the C#/VB impls of the language-neutral intrfaces, could then just bridge over to the compiler-specific APIs we were generating.  I would be amenable to this approach.

Reply:
> This also matters if we want to support things like LSIF for F#/TS/JS.

So this isn't something I expect to be doing at all: the LSIF generation code already speaks in terms of symbols and semantic models for performance reasons and how the model works. As F#/TS/JS move to LSP, they'll be much better equipped to reuse their code directly rather than having us as a middle proxy, since they'll also be able to better take advantage of however their symbol model/semantics model works. Consider @CyrusNajmabadi if we were to move the LSIF generator over to artifact production -- at this point that's not being used for F#/TS/JS either because they won't support that.

I haven't reviewed this PR yet so I won't speak to the rest of the layering concerns but wanted to call out that using our LSIF is a very much non-goal of the current design and approach, and isn't something that would really benefit anyone that I can see.

Reply:
>  since they'll also be able to better take advantage of however their symbol model/semantics model works

I woudl think they would still go through these APIs.  It seems bizarre that we literally have apis designed to allow languages to embed into an IDE.  And LSIF is just the encoding of that information to a file cache, but everyone would now go rewrite this.